keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resolvers/ImageResolver.kt

```diff
@@ -117,22 +90,35 @@ class ImageResolver(
     ) ?: throw NoImageFound(imageProvider.packageName)
 
     log.info("Image found for {}: {}", imageProvider.packageName, image)
-    return ResolvedImages(
-      image.appVersion,
-      image.toResolvedImage(resource.spec.locations.regions.map(SubnetAwareRegionSpec::region))
-    )
+    return image
   }
 
-  private fun NamedImage.toResolvedImage(regions: Collection<String>) =
-    amis
-      .filterKeys { it in regions }
-      .mapValues { (_, amis) -> amis?.first() }
+  private fun Resource<ClusterSpec>.withVirtualMachineImages(image: NamedImage): Resource<ClusterSpec> {
+    val imageIdByRegion = image
+      .amis
       .filterNotNullValues()
-      .also {
-        if (it.size < regions.size) {
-          throw NoImageFoundForRegions(imageName, regions - it.keys)
-        }
-      }
+      .filterValues { it.isNotEmpty() }
+      .mapValues { it.value.first() }
+    val missingRegions = spec.locations.regions.map { it.region } - imageIdByRegion.keys
+    if (missingRegions.isNotEmpty()) {
+      throw NoImageFoundForRegions(image.imageName, missingRegions)
+    }
+
+    val overrides = mutableMapOf<String, ServerGroupSpec>()
+    overrides.putAll(spec.overrides)
+    spec.locations.regions.map { it.region }.forEach { region ->
+      overrides[region] = overrides[region].withVirtualMachineImage(VirtualMachineImage(imageIdByRegion.getValue(region), image.appVersion))
+    }
+
+    return copy(spec = spec.copy(overrides = overrides))
+  }
+
+  private fun ServerGroupSpec?.withVirtualMachineImage(image: VirtualMachineImage) =
+    (this ?: ServerGroupSpec()).run {
```

Comment:
Sorry, another newbie question: can you help me read this? If `this` is `null` then run the block?! When would that ever happen?

Reply:
Yes, the receiver of the function is declared as nullable, so `this` can be `null` here. It's a pretty neat feature of Kotlin that you can do that. If the `?` were omitted from the function's receiver then the compiler would not allow this since it knows `this` could _not_ be `null`.

Reply:
Basically, in this specific case it's saying "use either `this` or a default new instance". This is because image ids are per region but it's probably unusual for a user to be overriding anything else in the server group spec on a per-region basis, so the `ServerGroupSpec` for any given region is likely not defined.

Reply:
I would be in favor of changing this function so it's a bit clearer, even if it's longer. The way this is written it's hard for me to understand as well.